ops/model.py

```diff
@@ -3141,19 +3142,21 @@ class _ModelBackend:
 
     def __init__(
         self,
+        juju_context: 'ops.main._JujuContext',
```

Comment:
@tonyandrewmeyer Will this signature change break ops-scenario? I don't really understand how this subclass can call `super().__init__()` without args here anyway: https://github.com/canonical/ops-scenario/blob/602f9b9ce0321ed383adc9f6de28047c3a4afac6/scenario/mocking.py#L101-L113

I'm not saying we shouldn't do it even if it does break Scenario, but we should think it through at any rate.

We could even make `juju_context` the final parameter, and make it optional, and if not provided this uses `_JujuContext` with `os.environ` by default. That would avoid lots of changes to our tests. Not as explicit, but seems a reasonable default.

Reply:
> @tonyandrewmeyer Will this signature change break ops-scenario?

It would break Scenario, yes. It wouldn't be too arduous to release a 6.x version that can handle it either way, and 7.0 could require ops 2.16 (it already needs ops 2.15).

> I don't really understand how this subclass can call `super().__init__()` without args here anyway: https://github.com/canonical/ops-scenario/blob/602f9b9ce0321ed383adc9f6de28047c3a4afac6/scenario/mocking.py#L101-L113

The other args have defaults :). I haven't checked, but I think that Scenario isn't passing those things through because it's handling the mocking in some other way.

Reply:
Yes, it's because other args are optional with defaults, I saw it in a bunch of tests where only unit_name is passed in.

Per the above discussion, I think @benhoyt's suggestion makes sense: I will make `juju_context` the final optional parameter and revert the test changes.

Reply:
Let's make it keyword-only argument then

Reply:
I think it probably ought to be keyword-only (there seems to be no logical order here) but note that will also break anything using this. It'd be a pretty straightforward update for Scenario, but this is a public class, even though we say not to instantiate it directly in the docstring.

Reply:
I've made it:

```python
def __init__(
    self,
    unit_name: Optional[str] = None,
    model_name: Optional[str] = None,
    model_uuid: Optional[str] = None,
    juju_context: Optional['ops.main._JujuContext'] = None,
):
```

and tested it, it did not break things. Making it as keyword-only also works, as in:

```python
def __init__(
    self,
    unit_name: Optional[str] = None,
    model_name: Optional[str] = None,
    model_uuid: Optional[str] = None,
    *,
    juju_context: Optional['ops.main._JujuContext'] = None,
):
```

But since the first version didn't break anything, I kept the first version for now. Feel free to reopen this conversation if I missed something.